genesis/src/lockups.rs

```diff
@@ -0,0 +1,160 @@
+//! lockups generator
+use solana_sdk::{clock::Epoch, epoch_schedule::EpochSchedule, timing::years_as_slots};
+use std::time::Duration;
+
+#[derive(Debug, Default, Clone)]
+pub struct Lockups {
+    /// where in iteration progress I am
+    expiration: usize,
+    /// number of expirations post cliff
+    expirations: usize,
+    /// fractionage offered to f() last time
+    prev_fraction: f64,
+    /// portion of total available at cliff
+    cliff_fraction: f64,
+    /// time of cliff, in epochs
+    cliff_epoch: Epoch,
+    /// number of post-cliff events
+    expiration_fraction: f64,
+    /// time between each post-cliff event
+    expiration_epochs: Epoch,
+}
+
+impl Lockups {
+    pub fn by_years(
+        cliff_fraction: f64,   // first cliff fraction
+        cliff_year: f64,       // first cliff time, in years
+        expirations: usize,    // number of follow-on expirations
+        expiration_years: f64, // years between each following expiration
+        epoch_schedule: &EpochSchedule,
+        tick_duration: &Duration,
+        ticks_per_slot: u64,
+    ) -> Self {
+        let cliff_slot = years_as_slots(cliff_year, tick_duration, ticks_per_slot) as u64;
+        let cliff_epoch = epoch_schedule.get_epoch(cliff_slot);
+
+        // assumes that the first cliff is after any epoch warmup
+        let first_expiration_slot =
+            years_as_slots(cliff_year + expiration_years, tick_duration, ticks_per_slot) as u64;
+        let expiration_epochs = epoch_schedule.get_epoch(first_expiration_slot) - cliff_epoch;
+
+        Self::by_epochs(cliff_fraction, cliff_epoch, expirations, expiration_epochs)
+    }
+
+    pub fn by_epochs(
+        cliff_fraction: f64,      // first cliff fraction
+        cliff_epoch: Epoch,       // first cliff epoch
+        expirations: usize,       //  number of follow-on expirations
+        expiration_epochs: Epoch, // epochs between each following expiration
+    ) -> Self {
+        let expiration_fraction = (1.0 - cliff_fraction) / expirations as f64;
+
+        Self {
+            prev_fraction: 0.0,
+            expiration: 0,
+            expirations,
+            cliff_fraction,
+            cliff_epoch,
+            expiration_fraction,
+            expiration_epochs,
+        }
+    }
+}
+
+impl Iterator for Lockups {
+    type Item = Expiration;
```

Comment:
Given the name `Lockups`, I would have expected `type Item = Lockup`.

Reply:
might rename Lockups to Schedule or Expirations, since Lockups doesn't carry anything else

Reply:
TokenReleaseSchedule?

Reply:
have "Unlocks -> Unlock" now...